dotnet/samples/KernelSyntaxExamples/Example58_FunctionCalling.cs

```diff
@@ -0,0 +1,104 @@
+
+        await CompleteChatWithFunctionsAsync("What day is today?", chatHistory, chatCompletion, kernel);
+
+        await CompleteChatWithFunctionsAsync("What computer tablets are available for under $200?", chatHistory, chatCompletion, kernel);
+    }
+
+    private static async Task<IKernel> InitializeKernelAsync()
+    {
+        // Create kernel with chat completions service
+        IKernel kernel = new KernelBuilder()
+            .WithLoggerFactory(ConsoleLogger.LoggerFactory)
+            .WithOpenAIChatCompletionService(TestConfiguration.OpenAI.ChatModelId, TestConfiguration.OpenAI.ApiKey, serviceId: "chat")
+            //.WithAzureChatCompletionService(TestConfiguration.AzureOpenAI.ChatDeploymentName, TestConfiguration.AzureOpenAI.Endpoint, TestConfiguration.AzureOpenAI.ApiKey, serviceId: "chat")
+            .Build();
+
+        // Load functions to kernel
+        string folder = RepoFiles.SampleSkillsPath();
+        kernel.ImportSemanticSkillFromDirectory(folder, "SummarizeSkill");
+        kernel.ImportSemanticSkillFromDirectory(folder, "WriterSkill");
+        kernel.ImportSemanticSkillFromDirectory(folder, "FunSkill");
+        kernel.ImportSkill(new TimeSkill(), "TimeSkill");
+
+        await kernel.ImportAIPluginAsync("KlarnaShoppingPlugin", new Uri("https://www.klarna.com/.well-known/ai-plugin.json"), new OpenApiSkillExecutionParameters());
+
+        return kernel;
+    }
+
+    private static async Task CompleteChatWithFunctionsAsync(string ask, ChatHistory chatHistory, IChatCompletion chatCompletion, IKernel kernel)
+    {
+        Console.WriteLine($"User message: {ask}");
+        chatHistory.AddUserMessage(ask);
+
+        // Retrieve available functions from the kernel and add to request settings
+        ChatRequestSettings requestSettings = new();
+        requestSettings.Functions = kernel.Skills.GetFunctionsView();
+
+        // Send request
+        var chatResult = (await chatCompletion.GetChatCompletionsAsync(chatHistory, requestSettings))[0];
+
+        // Check for message response
+        var chatMessage = await chatResult.GetChatMessageAsync();
+        if (chatMessage.Content.IsNotNullOrEmpty())
+        {
+            Console.WriteLine(chatMessage.Content);
+        }
+
+        // Check for function response
+        var functionCall = chatResult.ModelResult.GetResult<ChatModelResult>().Choice.Message.FunctionCall;
+        if (functionCall is not null)
+        {
+            FunctionCallResponse functionResponse = FunctionCallResponse.FromFunctionCall(functionCall);
+
+            Console.WriteLine("Function name: " + functionResponse.FunctionName);
+            Console.WriteLine("Skill name: " + functionResponse.SkillName);
+            Console.WriteLine("Arguments: ");
+
+            var context = kernel.CreateNewContext();
+            if (context.Skills!.TryGetFunction(functionResponse.SkillName, functionResponse.FunctionName, out ISKFunction? func))
```

Comment:
I believe that the logic for the AI function call should reside within SK (`SemanticFunction`?), so SK consumers do not need to implement it.

From the perspective of SK consumer code, it should be as simple as this:

1. Register/import a semantic function.
2. Register/import available native functions that might be called.
3. Execute the semantic function by calling `kernel.RunAsync`.
4. Retrieve the result returned by the `RunAsync` method.

There's no need to write any additional code. Everything (receiving the AI function response, finding/calling the function, calling AI again, receiving/parsing/returning the result) should be handled behind the scenes by SK.

Reply:
I suggested Gina go down this path as a first step and then we could build on as appropriate. SK treats this as ITextCompletion with regards to semantic functions today so a lot more changes would be needed.

Reply:
I think the logic you described makes sense in the scope of something like ActionPlanner which could leverage this functionality when supported by the available AIService

Reply:
In the discussions I've had with Shawn and others, there has been hesitance to automatically have SK call a function returned by the model without some sort of approval. (Mostly security and cost concerns) By leaving that step to the app, it gives the app more control over validating the function call. This is also how we currently handle planning - SK returns the plan, it's up to the app to execute it. I think it would be best to maintain that consistency in behavior.

However, we probably could consider moving the

Reply:
Here is an example of how this could be done as an extension: https://github.com/gitri-ms/semantic-kernel/pull/1

Reply:
I spoke with @shawncal and we have decided not to bring in that extension at this time. For now, with this PR, it will be up to the app to invoke the function. We have not written off the possibility of adding auto-execution in a future update though. I am making a note in the ADR and will open an issue to track follow-up on this.